src/classes/frPledge.cls

```diff
@@ -0,0 +1,88 @@
+*  1. Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*  2. Redistributions in binary form must reproduce the above copyright
+*     notice, this list of conditions and the following disclaimer in the
+*     documentation and/or other materials provided with the distribution.
+*  3. All advertising materials mentioning features or use of this software
+*     must display the following acknowledgement:
+*     This product includes software developed by the <organization>.
+*  4. Neither the name of the <organization> nor the
+*     names of its contributors may be used to endorse or promote products
+*     derived from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY FUNRAISE INC ''AS IS'' AND ANY
+*  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+*  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+*  DISCLAIMED. IN NO EVENT SHALL FUNRAISE INC BE LIABLE FOR ANY
+*  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+*  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+*  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+*  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+*
+*
+* PURPOSE:  A class for representing a Funraise pledge.  Used for creating & matching to pledges
+*
+*
+*
+* CREATED: 2020 Funraise Inc - https://funraise.io
+* AUTHOR: Alex Molina
+*/
+public class frPledge {
+    public static Pledge__c findActive(Id contactId) {
+        List<Pledge__c> activePledges = [SELECT Id FROM Pledge__c 
+                                         WHERE Supporter__c = :contactId 
+                                         AND (Start_Date__c = null OR Start_Date__c >= TODAY)
+                                         AND (End_Date__c = null OR End_Date__c <= TODAY)
+                                         AND Active__c = true
+                                         AND Pledge_Donation__c = null
+                                         ORDER BY CreatedDate ASC LIMIT 1
+                                        ];
+        return activePledges.size() > 0 ? activePledges.get(0) : null;
+    }
+    
+    public static Pledge__c create(Opportunity opp) {
+        Pledge__c pledge = new Pledge__c(
+            Name = frUtil.truncateToFieldLength(Pledge__c.Name.getDescribe(), opp.Name + ' Pledge'),
+            Supporter__c = opp.fr_Donor__c,
+            Pledge_Amount__c = opp.Amount,
+            Pledge_Donation__c = opp.Id,
+            Pledge_Donation_uq__c = opp.Id
+        );
+        try {
+            Database.upsert(pledge, Pledge__c.Pledge_Donation_uq__c , true);
```

Comment:
There are 2 fields on the pledge to track the entity that created the pledge.  Since either a pledge donation or a subscription can create a pledge in Salesforce, they are lookup fields to an opportunity or subscription in Salesforce respectively.  However, you are not allowed to mark lookup fields as unique OR as external id fields, so I made a second text field with a `_uq` suffix that should always have the same value as the lookup field.  This 2 fields accomplish two things:
1) Since they are marked as unique, that enforces the fact that a pledge donation or a subscription can only ever create one pledge
2) Since they are marked as external ids, we can do this upsert on them, which removes the need to query if it exists or needs an update in these requests.  By performing the upsert on this field, it will update the existing pledge record if it exists or create if it's missing